kep/real-load-aware-scheduling/real-load-aware-scheduling.md

```diff
@@ -0,0 +1,540 @@
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The 50% threshold value for utilisation will be made configurable.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+#### Spreading with BestFitBinPack
+
+
+To meet the 2nd constraint in the problem statement, we utilize the existing PodTopologySpread (PTS) plugin provided by K8s. It was proved with experiments that using a dynamic weight-based approach is not a good idea here. However, if we choose a fixed large weight for PTS, say 100, we can meet our requirements. The problem with this approach is that we will end up downplaying the scores of other important default plugins like Image Locality, Interpod Affinity, etc. with their default weights. Also, with more plugins we plan to add in the future, managing weights will not be simple. So for BFBP to work well with PTS without modifying weights, the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc.
+
+
+**Algorithm**
+
+
+
+1. For each node, calculate S = (1000*PTS + PTS*BFBP)
+2. Scale each S to be between 0-100
+3. Return the scores
+
+**Algorithm Analysis**
+
+1. The algorithm favors nodes with high PTS scores with 1000*PTS part. The weight 1000 is chosen to give preference to PTS  and to avoid the congestion of scores upon scaling down, my magnifying the scores 
+2. PTS*BFBP penalizes hot nodes with low BFBP scores.
+3. The final scaled-down value deals with congestion, hot node penalty, and favoring of high PTS score nodes
+
+Example: 
+
+
+| Node | PTS | BFBP | Combined Score |
+|------|-----|------|----------------|
+|  N1  | 100 |  35  |       96       |
+|  N2  |  67 |  75  |       67       |
+|  N3  | 100 |  80  |       100      |
+|  N4  |  82 |  90  |       83       |
+|  N5  |  0  |  90  |        0       |
+
+
+
+### Safe Balancing Plugin
+
+
+#### PreScore plugin
+
+This extension point, which is invoked for all the nodes at once, will be used to cache metrics for all nodes using a single REST API  call. This optimization will reduce multiple calls for each node from the Score plugin below, reducing overall scheduling latency.
```

Comment:
We did some performance benchmarks and PreScore was a bottleneck. This will be replaced with a background go routine that keeps updating metrics from load watcher.